File: Problems Solutions/56. Merge Intervals.py
# https://leetcode.com/problems/merge-intervals/

class Solution:
    def merge(self, intervals):
        intervals.sort()
        result = [intervals[0]]

        def merge(i1, i2):
            if i1[1] < i2[0]:
                return None
            
            if i1[0] <= i2[0] and i1[1] >= i2[1]:
                return i1
            
            return [i1[0], i2[1]]

        for i in range(1, len(intervals)):
            i1, i2 = result.pop(), intervals[i]
            merged = merge(i1, i2)

            if not merged:
                result.extend([i1, i2])
            else:
                result.append(merged)

        return result

tests_data = (
    ([[1,3],[2,6],[8,10],[15,18]], [[1,6],[8,10],[15,18]]),
    ([[1,4],[4,5]], [[1,5]])

)
obj = Solution()

for data in tests_data:
    expected = data[1]
    assert expected == obj.merge(data[0])

File: Problems Solutions/go/56. Merge Intervals.go
package main

import (
	"sort"
)

func _merge(i1 []int, i2 []int) [][]int {
    var r [][] int

    if i1[1] < i2[0] {
        r = append(append(r, i1), i2)
    } else if i1[0] <= i2[0] && i1[1] >= i2[1] {
        r = append(r, i1)
    } else if i2[0] <= i1[1] {
        var i = []int {i1[0], i2[1]} 
        r = append(r, i)
    }

    return r
}

func merge(intervals [][]int) [][]int {
    var r [][]int

    sort.Slice(
        intervals, func(i, j int) bool {
        return intervals[i][0] < intervals[j][0]
    })

    r = append(r, intervals[0])

    for _, i2 := range(intervals[1:]) {
        var i1 = r[len(r) - 1]
        var merged = _merge(i1, i2)

        // pop last element from result slice
        r = r[:len(r) - 1]

        for _, i := range(merged) {
            r = append(r, i)
        }
    }

    return r
}